library/testing/PseudoUnitTest.h

```diff
@@ -0,0 +1,150 @@
+#ifndef Test_h
+#define Test_h
+
+#include <functional>
+#include <iostream>
+#include <sstream>
+
+/** Helper to perform multiple checks within the same `ctest` test.
+ * Checks are performed using the various overloads of `operator()`
+ * and their results are logged and tracked so that the `result()` method
+ * will return `EXIT_FAILURE` if any of them failed (and `EXIT_SUCCESS` otherwise).
+ */
+class PseudoUnitTest
+{
+public:
+  /** test a boolean condition */
+  void operator()(const std::string& label, bool condition)
+  {
+    this->record(condition, label);
+  }
+
+  /** test the execution of a function */
+  void operator()(const std::string& label, std::function<void()> function)
+  {
+    this->testFunction<Dummy>(label, function);
+  }
+
+  /** test the execution of a function expecting a given exception */
+  template<typename E>
+  void expect(const std::string& label, std::function<void()> function)
+  {
+    this->testFunction<E>(label, function, true);
+  }
+
+  /** test the equality of two values with `==` */
+  template<typename T1, typename T2>
+  void operator()(const std::string& label, const T1& actual, const T2& expected)
+  {
+    const bool success = actual == expected;
+    this->record(success, label, this->comparisonMessage(actual, expected, success ? "==" : "!="));
+  }
+
+  /** test the equality of two values with a comparator */
+  template<typename T1, typename T2, typename F>
+  void operator()(const std::string& label, const T1& actual, const T2& expected, F comparator)
+  {
+    const bool success = comparator(actual, expected);
+    this->record(success, label, this->comparisonMessage(actual, expected, success ? "~=" : "!="));
+  }
+
+  int result()
+  {
+    const bool success = this->failCount == 0;
+    return success ? EXIT_SUCCESS : EXIT_FAILURE;
+  }
+
+private:
+  size_t failCount = 0;
+  size_t passCount = 0;
+
+  /* execute `function` and:
+   * - *fail* on normal termination when `expectException` is `true`
+   * - *pass* on normal termination when `expectException` is `false`
+   * - *pass* when an exception of type `E` is caught
+   * - *fail* when anything else is caught
+   */
+  template<typename E>
+  void testFunction(
+    const std::string& label, std::function<void()> function, bool expectException = false)
+  {
+    try
+    {
+      function();
+      if (expectException)
```

Comment:
Is it making sense to drop this boolean and use `if constexpr (std::is_same_v<E, Dummy>)` here?

Reply:
Way better, thanks!